### PR TITLE
Iot Hub Commandlets

### PR DESCRIPTION
**Pull request by rkmanda**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThrough` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@rkmanda__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (rmanda). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by rkmanda** [comment]
This is the next iteration after the older PR for the same was closed. Link to PR https://github.com/Azure/azure-powershell/pull/3156

**Comment by markcowl** [comment]
@rkmanda Please clean up the commits in this PR as described here:https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md

Please fill out the checklist as required as well

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+{
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.Management.IotHub.Common;
+    using Microsoft.Azure.Commands.Management.IotHub.Models;
+    using Microsoft.Azure.Management.IotHub;
+    using Microsoft.Azure.Management.IotHub.Models;
+    using ResourceProperties = Microsoft.Azure.Commands.Management.IotHub.Properties;
+
+    [Cmdlet(VerbsCommon.New, "AzureRmIotHubExportDevices", SupportsShouldProcess = true)]
+    [OutputType(typeof(JobResponse))]
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs]: PSIotHubJobResponse?

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            HelpMessage = "The BlobContainerUri to export to")]
+        [ValidateNotNullOrEmpty]
+        public string ExportBlobContainerUri { get; set; }
+
+        [Parameter(
+            Position = 3,
+            Mandatory = true,
+            HelpMessage = "ExcludeKeys")]
+        [ValidateNotNullOrEmpty]
+        public bool ExcludeKeys { get; set; }
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs]: Why not make this a switch parameter?

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubImportDevices.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+{
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.Management.IotHub.Common;
+    using Microsoft.Azure.Commands.Management.IotHub.Models;
+    using Microsoft.Azure.Management.IotHub;
+    using Microsoft.Azure.Management.IotHub.Models;
+    using ResourceProperties = Microsoft.Azure.Commands.Management.IotHub.Properties;
+
+    [Cmdlet(VerbsCommon.New, "AzureRmIotHubImportDevices", SupportsShouldProcess = true)]
+    [OutputType(typeof(JobResponse))]
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubImportDevices.cs]: PSIotHubJobResponse?

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHub.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                    // This is because the underlying IotHubClient sdk currently throws an exception when it receives a 404 during the 
+                    // long running delete operation. Remove this once the sdk is fixed to handle this.
+
+                    if (!e.Message.Contains("NotFound"))
+                    {
+                        throw e;
+                    }
+                }
+
+                this.WriteObject("Iot Hub Deleted");
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHub.cs]: Do not write text to the output stream.  In the case of delete, you should return nothing for success, and write an error for failure.  You should also implement a PassThrough parameter, and output True on successful delete if it is selected.

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            Position = 1,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Name of the Iot Hub")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(
+            Position = 2,
+            Mandatory = true,
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs]: Any chance this will come from the pipeline (A Get call, perhaps?)

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        [ValidateNotNullOrEmpty]
+        [ValidateSetAttribute(EventsEndpointName, OperationsMonitoringEventsEndpointName)]
+        public string EventHubEndpointName { get; set; }
+
+        [Parameter(
+            Position = 3,
+            Mandatory = true,
+            HelpMessage = "Name of the EventHub ConsumerGroupName")]
+        [ValidateNotNullOrEmpty]
+        public string EventHubConsumerGroupName { get; set; }
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs]: Any chance this will come from the pipeline (A Get call, perhaps?)

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubKey.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Name of the Iot Hub")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(
+            Position = 2,
+            Mandatory = true,
+            HelpMessage = "Name of the Key")]
+        [ValidateNotNullOrEmpty]
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubKey.cs]: Any chance this will come from the pipeline (A Get call, perhaps?)

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Management.Automation;
+    using System.Linq;
+    using Microsoft.Azure.Commands.Management.IotHub.Common;
+    using Microsoft.Azure.Commands.Management.IotHub.Models;
+    using Microsoft.Azure.Management.IotHub;
+    using Microsoft.Azure.Management.IotHub.Models;
+
+    [Cmdlet(VerbsCommon.Set, "AzureRmIotHub", SupportsShouldProcess = true)]
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs]: Need a default parameter set setting

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    [OutputType(typeof(PSIotHub))]
+    public class SetAzureRmIotHub : IotHubBaseCmdlet
+    {
+        const string UpdateSkuParameterSet = "UpdateSku";
+        const string UpdateEventHubEndpointPropertiesParameterSet = "UpdateEventHubEndpointProperties";
+        const string UpdateFileUploadPropertiesParameterSet = "UpdateFileUploadProperties";
+        const string UpdateCloudToDevicePropertiesParameterSet = "UpdateCloudToDeviceProperties";
+        const string UpdateOperationsMonitoringPropertiesParameterSet = "UpdateOperationsMonitoringProperties";
+
+        [Parameter(
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs]: If this is mandatory in all parameter sets, you can use a single parameter attribute with no ParameterSetName setting

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+            HelpMessage = "Name of the Resource Group")]
+        [Parameter(
+            ParameterSetName = UpdateOperationsMonitoringPropertiesParameterSet,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Name of the Resource Group")]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs]: Same comment

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+            ParameterSetName = UpdateFileUploadPropertiesParameterSet,
+            Mandatory = false,
+            HelpMessage = "fileUploadNotificationMaxDeliveryCount")]
+        [ValidateNotNullOrEmpty]
+        public int? FileUploadNotificationMaxDeliveryCount { get; set; }
+
+        [Parameter(
+            ParameterSetName = UpdateFileUploadPropertiesParameterSet,
+            Mandatory = true,
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs]: Any chance of any of these file settings coming from the pipeline?

File: src/ResourceManager/IotHub/Commands.IotHub/ResourceManagerStartup.ps1 [hunk 12]
```diff
@@ -0,0 +1,276 @@
+# ----------------------------------------------------------------------------------
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/ResourceManagerStartup.ps1]: What is this file intended for?  Remove

File: src/ResourceManager/IotHub/Commands.IotHub/help/Add-AzureRmIotHubEventHubConsumerGroup.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+```
+PS C:\> Add-AzureRmIotHubEventHubConsumerGroup -ResourceGroupName "myresourcegroup" -Name "myiothub" -EventHubEndpointName "operationsMonitoringEvents" -EventHubConsumerGroupName "myconsumergroup"
+```
+Adds a new consumergroup named "myconsumergroup" to the eventhub for operations monitoring events in the iothub named "myiothub"
+
+## PARAMETERS
+
+### -EventHubConsumerGroupName
+EventHubConsumerGroupName.
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/help/Add-AzureRmIotHubEventHubConsumerGroup.md]: Informative description, please

File: src/ResourceManager/IotHub/Commands.IotHub/help/Set-AzureRmIotHub.md [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+```
+PS C:\> Set-AzureRmIotHub -ResourceGroupName "myresourcegroup" -Name "myiothub" -EventHubRetentionTimeInDays 4
+```
+Update the retention time in days to 4 for both the telemetry and operationsmonitoringevents events for the IotHub named "myiothub"
+
+## PARAMETERS
+
+### -CloudToDevice
+CloudToDevice
+
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/help/Set-AzureRmIotHub.md]: Need an informative description for each parameter.  This applies to all help files

File: src/ResourceManager/IotHub/Commands.IotHub.Test/ScenarioTests/IotHubTests.ps1 [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+	$cloudToDevice = $iothubUpdated.Properties.CloudToDevice
+	$cloudToDevice.MaxDeliveryCount = 25
+	$iotHubUpdated = Set-AzureRmIotHub -ResourceGroupName $ResourceGroupName -Name $IotHubName -CloudToDevice $cloudToDevice
+	Assert-True { $iothubUpdated.Properties.CloudToDevice.MaxDeliveryCount -eq 25 }
+
+	# Operations Monitoring properties update
+	$op= $iotHubUpdated.Properties.OperationsMonitoringProperties
+	$op.OperationMonitoringEvents["Connections"] = "Information"
+	$iotHubUpdated = Set-AzureRmIotHub -ResourceGroupName $ResourceGroupName -Name $IotHubName -OperationsMonitoringProperties $op
+	Assert-True { $iothubUpdated.Properties.OperationsMonitoringProperties.OperationMonitoringEvents["Connections"] -eq "Information" }
```
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub.Test/ScenarioTests/IotHubTests.ps1]: Not as a part of the current PR. The test for this is a little involved as it requires a SAS token to be fetched on the fly. Will add the test later.

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/GetAzureRmIotHubJob.cs [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    using System.Collections.Generic;
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.Management.IotHub.Common;
+    using Microsoft.Azure.Commands.Management.IotHub.Models;
+    using Microsoft.Azure.Management.IotHub;
+    using Microsoft.Azure.Management.IotHub.Models;
+
+    [Cmdlet(VerbsCommon.Get, "AzureRmIotHubJob")]
+    [OutputType(typeof(PSIotHubJobResponse), typeof(List<PSIotHubJobResponse>))]
+    public class GetAzureRmIotHubJob : IotHubBaseCmdlet
```
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/GetAzureRmIotHubJob.cs]: The prereq for testing this is to have a test for NewAzureRmIotHubExportDevices  which also relies on a valid SAS token to be generated on the fly. Will target this later.

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.Management.IotHub.Common;
+    using Microsoft.Azure.Commands.Management.IotHub.Models;
+    using Microsoft.Azure.Management.IotHub;
+    using Microsoft.Azure.Management.IotHub.Models;
+    using ResourceProperties = Microsoft.Azure.Commands.Management.IotHub.Properties;
+
+    [Cmdlet(VerbsCommon.New, "AzureRmIotHubExportDevices", SupportsShouldProcess = true)]
+    [OutputType(typeof(JobResponse))]
+    public class NewAzureRmIotHubExportDevices : IotHubBaseCmdlet
```
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs]: Relies on a valid SAS token to be generated on the fly. Will target this later.

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubImportDevices.cs [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    public class NewAzureRmIotHubImportDevices : IotHubBaseCmdlet
```
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubImportDevices.cs]: Relies on a valid SAS token to be generated on the fly. Will target this later.

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/GetAzureRmIotHub.cs [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.Management.IotHub.Common;
+    using Microsoft.Azure.Commands.Management.IotHub.Models;
+    using Microsoft.Azure.Management.IotHub;
+    using Microsoft.Azure.Management.IotHub.Models;
+    using Microsoft.Rest.Azure;
+
+    [Cmdlet(VerbsCommon.Get, "AzureRmIotHub", DefaultParameterSetName = "ListIotHubsBySubscription")]
```
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/GetAzureRmIotHub.cs]: left by mistake

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs (hunk 5 above)
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs]: Nope

File: src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs (hunk 11 above)
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs]: Nope

File: src/ResourceManager/IotHub/Commands.IotHub.Test/ScenarioTests/IotHubTests.cs [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            XunitTracingInterceptor.AddToContext(new XunitTracingInterceptor(output));
+        }
+
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void TestAzureIotHubLifeCycle()
+        {
+            IotHubController.NewInstance.RunPsTestWorkflow(
+                () => { return new[] { string.Format("{0} {1} {2} {3} {4}", "Test-AzureRmIotHubLifecycle", "northeurope", "powershelliothub", "powershellrg", "S1") }; },
```
Comment (rkmanda) [inline src/ResourceManager/IotHub/Commands.IotHub.Test/ScenarioTests/IotHubTests.cs]: Found other resources hardcoding location directly in the script. What I have done is to ensure that the script itself does not have the hardcoding so it can be run in any location. Let me know if this works for the current PR. I will raise an issue to fix this in a subsequent PR.

**Comment by rkmanda** [comment]
@markcowl please review. Addressed most comments. Some tests are to be added and I will raise an issue for this. The region hardcoding will also be removed as a part of the test issue.

File: src/ResourceManager/IotHub/ChangeLog.md [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+        - Additional information about change #4
+
+    ## YYYY.MM.DD - Version X.Y.Z (Previous Release)
+    * Overview of change #1
+        - Additional information about change #1
+-->
+## Current Release
+
+## Version 1.0.0
+* Adds commandlets for the Azure IoT Hub
```
Comment (markcowl) [inline src/ResourceManager/IotHub/ChangeLog.md]: Please listy the capabilities of these cmdlets, and the resources they allow you to manage.

File: src/ResourceManager/IotHub/Commands.IotHub/Commands.IotHub.csproj [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>..\..\..\Package\Release\ResourceManager\AzureResourceManager\AzureRM.IotHub</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
```
Comment (markcowl) [inline src/ResourceManager/IotHub/Commands.IotHub/Commands.IotHub.csproj]: Why doi you have two property groups for the release configuration?  Please clean up.

**Comment by markcowl** [comment]
on demand run here:http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1339/

signing run here: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/746/

**Comment by cormacpayne** [comment]
Signing run: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/753/

**Comment by cormacpayne** [comment]
Sign job was successful: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/755/

**Comment by cormacpayne** [comment]
On demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1346/

Sign job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/760/

LGTM once both pass

**Comment by rkmanda** [comment]
Can this be merged now?
